docs/src/main/asciidoc/config-reference.adoc

```diff
@@ -358,6 +358,7 @@ quarkus.datasource.jdbc.url=jdbc:mysql://${application.server}:3306/mydatabase?u
 Additionally, the Expression Expansion engine supports the following segments:
 
 * `${expression:value}` - Provides a default value after the `:` if the expansion doesn't find a value.
+* `${expression:}` - Provides a `null` value if the expansion doesn't find a value.
```

Comment:
This is not completely correct. The provided value is actually an empty String. What happens later is that because empty Strings are converted to `null`, we get the `java.util.NoSuchElementException`. For instance, something like  `${a:a}${b:}${c:c}` returns `ac`.

Reply:
Interesting. My specific use case was that I wanted to be able to start my program with an environment variable undefined like so.

```
// application.properties
// Use BAR from env, or default to null if it's undefined
foo.bar=${BAR:}

// MyClass.java
// Will be null if BAR is undefined in env
@ConfigProperty(name = "foo.bar")
String optionalVar;
```

This works, but perhaps this is a smell/side-effect and shouldn't be encouraged? 

Of course, an alternative is to provide a default, but it doesn't necessarily make sense in all scenarios:

```
// application.properties
// Use BAR from env
foo.bar=${BAR}

// MyClass.java
// Will be "default-bar-value" if BAR is undefined in env. Omitting defaultValue will cause an expansion error
@ConfigProperty(name = "foo.bar", defaultValue = "default-bar-value")
String optionalVar;
```

Reply:
Yes, setting an empty String, clears the configuration value, but this is because of the internal conversion rules. Generally, this is not an issue. In this particular case, because all property expansion is evaluated before conversion rules take place, you get an empty string and not a null.

Reply:
Apologies. My example had an error since it has been a while since I first encountered this. 

It should have been:

```
@ConfigProperty(name = "foo.bar")
Optional<String> optionalVar;

// check optionalVar.isPresent(); later in code
```

The above code will not run if I do `foo.bar=${BAR}`. I found that it will run if I do `foo.bar=${BAR:}`. Quarkus docs state:

> If an expression cannot be expanded and no default is supplied a `NoSuchElementException` is thrown

So it sounds like optional vars are not really supported, and `foo.bar=${BAR:}` working is just a hacky workaround? If so I guess we can close this 😄